privacyidea/lib/tokens/tantoken.py

```diff
@@ -0,0 +1,218 @@
+"""
+
+import logging
+from privacyidea.lib.log import log_with
+from privacyidea.lib.tokenclass import TokenClass
+from privacyidea.lib.tokens.papertoken import PaperTokenClass
+from privacyidea.lib.policy import SCOPE
+from privacyidea.lib import _
+from privacyidea.lib.policydecorators import libpolicy
+from privacyidea.lib.crypto import geturandom, hash
+import binascii
+
+log = logging.getLogger(__name__)
+DEFAULT_COUNT = 100
+SALT_LENGTH = 4
+
+
+class TANACTION(object):
+    TANTOKEN_COUNT = "tantoken_count"
+
+
+class TanTokenClass(PaperTokenClass):
+    """
+    The TAN token allows to print out the next e.g. 100 OTP values.
+    This sheet of paper can be used to authenticate and strike out the used
+    OTP values. In contrast to the paper token, the OTP values of the TAN token
+    can be used randomly.
+    """
+
+    @log_with(log)
+    def __init__(self, db_token):
+        """
+        This creates a new TAN token object from a DB token object.
+
+        :param db_token: instance of the orm db object
+        :type db_token:  orm object
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"tan")
+        self.hKeyRequired = False
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the token type shortname
+
+        :return: 'paper'
+        :rtype: string
+        """
+        return "tan"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        Return the prefix, that is used as a prefix for the serial numbers.
+        :return: PITN
+        """
+        return "PITN"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        returns a subtree of the token definition
+
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or scalar
+        """
+        res = {'type': 'tan',
+               'title': 'TAN Token',
+               'description': 'TAN: TANs printed on a sheet '
+                              'of paper.',
+               'init': {},
+               'config': {},
+               'user':  ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       TANACTION.TANTOKEN_COUNT: {
+                           "type": "int",
+                           "desc": _("The number of OTP values, which are "
+                                     "printed on the paper.")
+                       }
+                   }
+               }
+               }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+        return ret
+
+    def update(self, param, reset_failcount=True):
+        tan_dict = {}
```

Comment:
As ``tan_dict`` is defined in both branches, we can omit this line.

Reply:
You are right. On the other hand the two if branches are very long! If we change them one day, we might end up with an undefined dict.
What do you think should be the best practice here? Maybe you are right...

Reply:
Hmm, I would think that in that case, having a undefined dict would be better than having a dictionary which is unintentionally empty, because the undefined dict would get noticed faster :)
However, this is probably a bit nit-picky. So I would be okay with the above extra line as well.

Reply:
Done.